### PR TITLE
Added support for using IServiceProvider when configuring URI and options in Azure Key vault health check

### PR DESCRIPTION
**Pull request by sanderaernouts**
**What this PR does / why we need it**:
This pull request adds support for configuring the key vault  URI and option using an `IServiceProvider` similar to how the `SqlServerHealthCheck` supports this for a connection string. This way the ASP.net core options pattern can be used to retrieve the key vault URI and/or keys and secrets.

So instead of this:
```csharp
var provider = services.BuildServiceProvider();
var settings = services.GetRequiredService<IOptions<MyOptions>>();

services.AddHealthChecks().AddAzureKeyVault(
    settings.Value.KeyVaultUri, 
    ..., 
    setup=> setup.AddSecret(settings.Value.MySecret))
```

You can now do something like this:
```csharp
services.AddHealthChecks().AddAzureKeyVault(
    sp => services.GetRequiredService<IOptions<MyOptions>>().Value.KeyVaultUri,
    ..., 
    (sp, setup) => setup.AddSecret(services.GetRequiredService<IOptions<MyOptions>>().Value.MySecret));
```

**Which issue(s) this PR fixes**:
*none*

**Special notes for your reviewer**:
This is based on the `connectionStringFactory` implementation in the `SqlServerHealthCheck`.

**Does this PR introduce a user-facing change?**:
It extends but does not break the existing options for configuring a `AzureKeyVaultHealthCheck`

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [ ] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs [hunk 1]
```diff
@@ -28,15 +28,67 @@ public static class AzureKeyVaultHealthChecksBuilderExtensions
...
+            }
+
+            return AddAzureKeyVault(builder, _ => keyVaultServiceUri, credential, (_, options) => setup?.Invoke(options), name, failureStatus, tags, timeout);
+        }
+
+        /// <summary>
+        /// Add a health check for Azure Key Vault. Default behaviour is using Managed Service Identity, to use Client Secrets call UseClientSecrets in setup action
+        /// </summary>
+        /// <param name="keyVaultServiceUri">The AzureKeyVault service uri.</param>
+        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzureCredential or other kind of TokenCredential,you can read more on <see href="https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md"/>. </param>
```
Comment (sungam3r) [inline src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs:43]:
```suggestion
        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzureCredential or other kind of TokenCredential, you can read more on <see href="https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md"/>. </param>

```

File: src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs [hunk 2]
```diff
@@ -28,15 +28,67 @@ public static class AzureKeyVaultHealthChecksBuilderExtensions
...
+
+            return AddAzureKeyVault(builder, _ => keyVaultServiceUri, credential, setup, name, failureStatus, tags, timeout);
+        }
 
-            if (keyVaultServiceUri == null) throw new ArgumentNullException(nameof(keyVaultServiceUri));
+        /// <summary>
+        /// Add a health check for Azure Key Vault. Default behaviour is using Managed Service Identity, to use Client Secrets call UseClientSecrets in setup action
+        /// </summary>
+        /// <param name="keyVaultServiceUriFactory">A factory to build the key vault URI to use.</param>
+        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzureCredential or other kind of TokenCredential,you can read more on <see href="https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md"/>. </param>
```
Comment (sungam3r) [inline src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs:68]:
```suggestion
        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzureCredential or other kind of TokenCredential, you can read more on <see href="https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md"/>. </param>

```

**Comment by Meertman** [comment]
Is it also not an option to look into the use of the Microsoft.Extensions.Azure library? (See here for documentation: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/microsoft.extensions.azure-readme-pre#getting-started)

This allows the configuration of the clients to use in a central place when configuring your services.

File: src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs (hunk 1 above)
Comment (sanderaernouts) [reply src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs]: Resolved your comment

File: src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs (hunk 2 above)
Comment (sanderaernouts) [reply src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs]: Resolved your comment

**Comment by sanderaernouts** [comment]
> 
> 
> Is it also not an option to look into the use of the Microsoft.Extensions.Azure library? (See here for documentation: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/microsoft.extensions.azure-readme-pre#getting-started)
> 
> This allows the configuration of the clients to use in a central place when configuring your services.

@Meertman that would be an option, but that means changing the implementation of the `AzureKeyVaultHealthCheck` significantly it seems. Don't think I want to drag that into this PR.

**Comment by sanderaernouts** [comment]
@sungam3r can this be merged?

**Comment by sungam3r** [comment]
I'm not an owner of this repo and have no write access to merge PRs.

**Comment by sanderaernouts** [comment]
Apologies, assumed you were because you reviewed the PR. @unaizorrilla can you have a look?

**Comment by unaizorrilla** [comment]
Yeap, I need to merge most of the PR on master before move to NET 6 :-( ...

@sungam3r  are you interested on colaborate as maintainer on this project? If is true, please let me know on unai at plainconcepts dot com

**Comment by unaizorrilla** [comment]
Merged, but, some details here!


 - The PR contains a file  "Modified' on SqlServerExtensions ( with no changes, probably line ending or something similar), for the futures PR please ensure only necessary files are 'touched'

Thanks for the PR a new release build is on app veyord , when finished a package 5.0.3 will be available on nuget.

**Pull request merged**